src/terminal/terminal.rs

```diff
@@ -545,51 +580,162 @@ where
             return Ok(());
         }
 
-        // Clear the viewport off the screen
-        self.clear()?;
-
-        // Move the viewport by height, but don't move it past the bottom of the terminal
-        let viewport_at_bottom = self.last_known_area.bottom() - self.viewport_area.height;
-        self.set_viewport_area(Rect {
-            y: self
-                .viewport_area
-                .y
-                .saturating_add(height)
-                .min(viewport_at_bottom),
-            ..self.viewport_area
-        });
-
-        // Draw contents into buffer
+        // Draw contents to insert into a buffer. The rest of this function is concerned with
+        // drawing this buffer onto the screen.
         let area = Rect {
-            x: self.viewport_area.left(),
+            x: 0,
             y: 0,
             width: self.viewport_area.width,
             height,
         };
         let mut buffer = Buffer::empty(area);
         draw_fn(&mut buffer);
-
-        // Split buffer into screen-sized chunks and draw
-        let max_chunk_size = (self.viewport_area.top() as usize) * (area.width as usize);
-        for buffer_content_chunk in buffer.content.chunks(max_chunk_size) {
-            let chunk_size = (buffer_content_chunk.len() / (area.width as usize)) as u16;
-
-            self.backend
-                .append_lines(self.viewport_area.height.saturating_sub(1) + chunk_size)?;
-
-            let iter = buffer_content_chunk.iter().enumerate().map(|(i, c)| {
-                let (x, y) = buffer.pos_of(i);
-                (
-                    x,
-                    self.viewport_area.top().saturating_sub(chunk_size) + y,
-                    c,
-                )
-            });
-            self.backend.draw(iter)?;
-            self.backend.flush()?;
-            self.set_cursor_position(self.viewport_area.as_position())?;
+        let mut buffer = buffer.content.as_slice();
+
+        // Use i32 variables so we don't have worry about overflowed u16s when adding, or about
+        // negative results when subtracting.
+        let mut drawn_height: i32 = self.viewport_area.top().into();
+        let mut buffer_height: i32 = height.into();
+        let viewport_height: i32 = self.viewport_area.height.into();
+        let screen_height: i32 = self.last_known_area.height.into();
+
+        while buffer_height + viewport_height > screen_height {
+            // There isn't enough room on the screen for the whole buffer and the viewport. We will
```

Comment:
I appreciate this very detailed and explanatory comment. However, it feels a bit too much, should we strip some parts or simplify it maybe? Any thoughts @joshka @kdheepak?

Reply:
It is quite the wall of text. I'm happy to remove some or all of it.

It took me a bit of time to figure out what was there before, so I was hoping to save somebody in the future a similar problem when trying to figure out the magic formula I left behind. But maybe the comment is so long and hard to follow, that it's as hard or harder to follow than it would be to just try to figure out the magic formula.

Just let me know what you want me to do with the comment.

Reply:
One approach could be to take out the proof. The comment could just end after the four constraints and the explanation of why they're important. An interested reader could then prove to themselves that the chosen `scroll_up` satisfied all four constraints, without being walked through it step-by-step.

Reply:
Yes, that makes sense to me. You can simply add the proof to the related issue or this PR and link it in the comment for those who wants to read more about it.

Reply:
I have moved the proof to a comment in this PR, and changed the code's comment accordingly.

Reply:
By moving around the code so that the top level method appears before the code which it calls, you might find that the detail you have here naturally gravitates into the doc comments of the private functions that you're calling instead of being in a single place here.

The main problem I see with this comment and the one above this is that it dives into the detail a bit too fast without setting up the context. This is likely a good amount of detail about things you found out while implementing, but without having gone through that process with you it seems like it's not tethered to an understanding of what this code is supposed to be doing. Starting with the assumption that the reader of this knows nothing about what's going on here (this is reasonably close to correct for me at least ;)), try to explain the algorithm / concerns from the outside in rather than the inside out.

Reply:
I rewrote the comments in the function. Please take a look and tell me what you think.